infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SnapshotSender.java

```diff
@@ -58,123 +74,159 @@
+                snapshotReadMessage = snapshotReader.read(snapshotSyncEventId);
+
+                // Mark the readingCompleted if there is no more messages.
+                readingCompleted = snapshotReadMessage.isEndRead();
+
+            } catch (TrimmedException te) {
+                log.warn("Cancel snapshot sync due to trimmed exception.", te);
+                dataSenderBufferManager.reset(Address.NON_ADDRESS);
+                snapshotSyncCancel(snapshotSyncEventId, LogReplicationError.TRIM_SNAPSHOT_SYNC);
+                cancel = true;
+                break;
+
+            } catch (Exception e) {
+                log.error("Caught exception during snapshot sync", e);
+                snapshotSyncCancel(snapshotSyncEventId, LogReplicationError.UNKNOWN);
+                cancel = true;
+                break;
+            }
+
+            messagesSent += processReads(snapshotReadMessage.getMessages(), snapshotSyncEventId);
+            observedCounter.setValue(messagesSent);
+        }
+
+       if (!cancel && !stopSnapshotSync) {
+            // Maximum number of batch messages sent. This snapshot sync needs to continue.
+            // Snapshot Sync is not performed in a single run, as for the case of multi-site replication
+            // the shared thread pool could be lower than the number of sites, so we assign resources in
+            // a round robin fashion.
+            log.trace("Snapshot sync continue for {} on timestamp {}", snapshotSyncEventId, baseSnapshotTimestamp);
+            fsm.input(new LogReplicationEvent(LogReplicationEventType.SNAPSHOT_SYNC_CONTINUE,
+                    new LogReplicationEventMetadata(snapshotSyncEventId)));
+       } else {
+           log.warn("Snapshot sync stopped for {} on timestamp {}", snapshotSyncEventId, baseSnapshotTimestamp);
+       }
+    }
+
+    /**
+     *
+     * @param snapshotSyncEventId
+     */
+    public void querySnapshotSyncStatus(UUID snapshotSyncEventId) {
+        if (!readingCompleted) {
+            log.error("It is in the wrong state, the reading is not in the complete state {}", readingCompleted);
+            return;
+        }
+
+        try {
+            // Query receiver status
+            LogReplicationQueryMetadataResponse response = dataSender.sendQueryMetadataRequest().get();
+
+            // If it has finised applying the snapshot, transition to log entry sync
+            // Otherwise query the status in another cycle.
+            if (response.getLastLogProcessed() == response.getSnapshotApplied() &&
+                response.getSnapshotApplied() == baseSnapshotTimestamp) {
+                log.info("SNAPSHOT full sync complete according to response {}", response);
+                snapshotSyncComplete(snapshotSyncEventId);
+            } else {
+                log.info("The receiver hasn't finished applying the snapshot yet {}", response);
+                fsm.input(new LogReplicationEvent(LogReplicationEventType.SNAPSHOT_SYNC_CONTINUE,
```

Comment:
I would address this issue in another PR. I really don't want to rebase any more. If you think this PR is not valuable, I can trash it from now on.

Reply:
No but this comment is not related to the sleep time, that is a different comment. In this one I was trying to convey to the idea that not having a state to transition to while waiting, but staying on the same one is not correct and I'm not sure the secondary effect as we re-enter in snapshot sync state which on entry has the behavior of triggering snapshot sync (which already happened). If you want I can help inserting a new state to transition on the WAIT_FOR_COMPLETE and leave the place holder so you move the code there and then the FSM is clean. If you're ok with that, I can push a commit on this PR with that minor change so we push the snapshot sync in phases in its complete form in this PR.

Reply:
I would prefer to push this PR in first. And you make new PR to amend it. As this PR has changes over a lot files, it is easily to have conflicts with other changes and it is hard to make the logic correct while resolving the conflicts. It doesn't break any function tests. If it just changes one or two files, I don't mind to rebase  one more time.

Reply:
I don't think it is a good idea to push something incomplete, specially if we're almost there. But if so, please take ownership of this item. I'll tell you how it would break:

- Let's say we finished snapshot transfer at time X. And from time X ---> Y we are applying. 
- Suppose active cluster is restarted or even simpler leadership changes in between this time (X---->Y) on the active. Re-negotiation would occur to determine next step. And negotiation would tell us, we're in the middle of a SNAPSHOT_APPLY. 
- Because we have no state to represent this, and IN_SNAPSHOT_SYNC state is being reused. We would move to that state, it generates a new event ID, and it starts a new snapshot sync. 
- The receiver will cancel the ongoing apply and clear everything. And start all over again.
- So, the purpose of optimizing because the receiver was APPLYING is lost. 
- If leadership changes again, we lose and start all over again. 
Therefore, we are not resilient to leadership changes either in the remote or local, when in snapshot sync. This would only work if there are no leadership changes.

Reply:
What happen to the current implementation?  Also the current implementation for any applying phase longer than 5 seconds it will go into the infinite loop for snapshot full sync.
I just tried to solve one thing at one time. 
I say this change is the frame work for further optimization.

Reply:
mm no, right? Cause in the current implementation we transition right away to log entry sync when the transfer is completed. We are trying to introduce the II phase. Which looking at the code made me realize that probably there is something else that is also missing.

Look at lines 106-130 in the InSnapshotSyncState for the current PR. When the ACK for the transfer is received (SNAPSHOT_SYNC_COMPLETE) which is inserted by LogReplicationSourceManager, we cancel the ongoing snapshot sync and transition to LOG_ENTRY (even if the apply is still going on). So, If we didn't reach the sleep line before the ACK is received, we just transition. And if we did hit the sleep we will block the thread and then when sleep time expires we will query for the status, leaving us in the following:
- If it had completed, we will insert a SNAPSHOT_SYNC_COMPLETED which will be ignored as we had already transitioned to IN_LOG_ENTRY
- If it failed, we will insert a SNAPHOT_SYNC_CONTINUE which will also be ignored as we already transitioned to IN_LOG_ENTRY. 

I think we are missing a couple of things to really get the II phase working, currently that sleep is not doing much,  as it will already have transitioned and regardless of the result we are already somewhere else.

Reply:
You are totally off the point for the leadership change. So if there is a leadership change for the current implementation, the new active leader will do the negotiation and start the full snapshot sync too. 
You are also missing the point that for any full snapshot sync with applying phase that takes longer than 5 seconds, our current implementation will go into infinite loop.

Reply:
I never said, that leadership does not start full snapshot sync on current implementation. Effectively it does, and that's one of the nice things about bringing in II phase, that we can avoid that and optimize. My point was precisely that we still don't actually have that, and that should be part of the II phase outcomes.

On the second point, can you kindly explain it to me. Cause I don't see that (just for my understanding, because either way we should look into how to improve and fix that with our new 2-phased approach).

Reply:
When sender  sends the snapshot_end message, it will block on the ack, and all completedFuture will be removed or timeout in 5 seconds. That is why we should not rely on ack for the completion.

Reply:
At least for netty, I remember for each completedFuture, it has a timeout of 5 seconds, then it will be removed. I am not sure if you have changed that part or not. 
Also for the snapshot sender, it depends on a timeout value. 
 LogReplicationEntry ack = snapshotSyncAck.get(DEFAULT_TIMEOUT, TimeUnit.MILLISECONDS);